source/extensions/filters/network/mysql_proxy/mysql_utils.h

```diff
@@ -51,6 +51,71 @@ class BufferHelper : public Logger::Loggable<Logger::Id::filter> {
   static DecodeStatus peekHdr(Buffer::Instance& buffer, uint32_t& len, uint8_t& seq);
 };
 
+/**
+ * MySQL auth method.
+ */
+enum class AuthMethod : uint8_t {
+  Unknown,
+  OldPassword,
+  NativePassword,
+  Sha256Password,
+  CacheSha2Password,
+  ClearPassword
+};
+
+// Helper function collection of old_password
+struct OldPassword {
+public:
+  static std::vector<uint8_t> generateSeed();
+  static std::vector<uint8_t> signature(const std::string& password,
+                                        const std::vector<uint8_t>& seed);
+  static std::vector<uint32_t> hash(const std::string& text) {
+    return hash(text.data(), text.size());
+  }
+  static std::vector<uint32_t> hash(const std::vector<uint8_t>& text) {
+    return hash(reinterpret_cast<const char*>(text.data()), text.size());
+  }
+  /*
+   * Generate binary hash from raw text string
+   * Used for Pre-4.1 password handling
+   */
+  static std::vector<uint32_t> hash(const char* text, int size);
+  static constexpr int SEED_LENGTH = 8;
+
+private:
+  struct RandStruct {
+    RandStruct(uint32_t seed1, uint32_t seed2);
+    double myRnd();
+    uint32_t seed1_, seed2_, max_value_;
+    double max_value_dbl_;
+  };
+};
+
+// Helper function collection of native_password
+struct NativePassword {
+  static std::vector<uint8_t> generateSeed();
+  static std::vector<uint8_t> signature(const std::string& password,
+                                        const std::vector<uint8_t>& seed);
+
+  static std::vector<uint8_t> hash(const std::string& text) {
+    return hash(text.data(), text.size());
+  }
+  static std::vector<uint8_t> hash(const std::vector<uint8_t>& text) {
```

Comment:
Lots of methods in _Native_ and _Old_ passwords overlap. Can you create a base class call _Password_ and derive _NativePassword_ and _OldPassword_ from _Password_? I know you want methods to be static, but it is still possible to use inheritance. Define _Password::generateSeed(length)_ and in derived classes call _generateSeed(20)_ for example.

Reply:
I'm trying to avoid construct a helper class before using these functions, so just use the template to implement it.